ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -274,56 +278,88 @@ initStartingWith tracer conf streamAPI initDb = do
 -- testing purposes, 'takeSnapshot' returns the block reference corresponding
 -- to the snapshot that we wrote.
 --
--- NOTE: This is a lower-level API that unconditionally takes a snapshot
--- (i.e., independent from whether this snapshot corresponds to a state that
--- is more than @k@ back).
+-- NOTE: This is a lower-level API that takes a snapshot independent from
+-- whether this snapshot corresponds to a state that is more than @k@ back. It
+-- doesn't take a snapshot if a snapshot at the same slot already exists on
+-- disk or the tip is the genesis.
+--
+-- Note that an EBB can have the same slot number as the block after it. This
+-- doesn't matter. The one block difference in the ledger state doesn't warrant
+-- an additional snapshot. The slot number in the name of the snapshot is only
+-- indicative, we don't rely on it being correct.
 --
 -- TODO: Should we delete the file if an error occurs during writing?
 takeSnapshot :: forall m l r h. MonadThrow m
              => Tracer m (TraceEvent r)
              -> HasFS m h
              -> (l -> Encoding)
              -> (r -> Encoding)
-             -> LedgerDB l r -> m (DiskSnapshot, WithOrigin r)
-takeSnapshot tracer hasFS encLedger encRef db = do
-    ss <- nextAvailable <$> listSnapshots hasFS
-    writeSnapshot hasFS encLedger encRef ss oldest
-    traceWith tracer $ TookSnapshot ss (csTip oldest)
-    return (ss, csTip oldest)
+             -> (r -> DiskSnapshot)
+             -> LedgerDB l r
+             -> m (Maybe DiskSnapshot, WithOrigin r)
+takeSnapshot tracer hasFS encLedger encRef toSnap db = case tip of
+    Origin      -> return (Nothing, tip)
+    NotOrigin r -> do
+      let snapShot = toSnap r
+      snaps <- listSnapshots hasFS
+      if List.any (sameSlot snapShot) snaps
+        then return (Nothing, tip)
+        else do
+          writeSnapshot hasFS encLedger encRef snapShot oldest
+          traceWith tracer $ TookSnapshot snapShot tip
+          return (Just snapShot, tip)
   where
     oldest :: ChainSummary l r
     oldest = ledgerDbAnchor db
 
+    tip = csTip oldest
+
 -- | Trim the number of on disk snapshots so that at most 'onDiskNumSnapshots'
--- snapshots are stored on disk. The oldest snapshots are deleted.
---
--- The deleted snapshots are returned.
+-- snapshots are stored on disk. The oldest snapshots are deleted. Snapshots
+-- with a suffix are not deleted.
 trimSnapshots :: Monad m
               => Tracer m (TraceEvent r)
               -> HasFS m h
               -> DiskPolicy
-              -> m [DiskSnapshot]
+              -> m ()
 trimSnapshots tracer hasFS DiskPolicy{..} = do
     snapshots <- listSnapshots hasFS
+    -- We ignore named snaspshots
+    let snapshots' = filter (isNothing . dsSuffix) snapshots
```

Comment:
```suggestion
    -- We ignore named snapshots
    snapshots <- filter (isNothing . dsSuffix) <$> listSnapshots hasFS
```
Now we can't accidentally refer to `snapshots` instead of `snapshots'`.